### DISCUSSION TITLE
when to use ISession vs IDocumentManager?

### DISCUSSION THREAD

**Discussion by weirdyang** [OrchardCMS/OrchardCore#16433] [Q&A] [answered]
I'm looking at the code for Queries and Workflows, and I noticed that Workflow module creates indexes and uses ISession to save the workflows and workflow types, where as the Queries module saves a single QueryDocument, when should I use the different approaches? How is the workflow document saved?

**Comment by Piedone** [comment, marked as answer]
`ISession` is the low-level data access service where you essentially interact with SQL in the YesSql model. Whatever you want to do in the document DB, you can do it via `ISession`. And if you want to fetch documents via arbitrary queries (i.e. `SELECT ... WHERE`s) you have to do it via `ISession`.

`IDocumentManager` is a higher-level services that uses `ISession` and provides a simplified, as well as cached access layer to single documents of a given type. This is useful if for your feature you want to keep a single piece of JSON in the DB. E.g. Queries keeps all the Queries in a single document, because you'll never really have more than a handful of them, so having them in separate documents (i.e. DB rows) would hurt performance.

Workflows takes a different approach (whether warranted or not) with storing all `WorkflowType` documents separately. Since `IDocumentManager` is only about document types that'll only ever have a single instance, it can't use it.

**Reply by weirdyang** [reply]
Thanks,
